### PR TITLE
Make sequential poromechanics to conserve mass

### LINKED ISSUE #2845: Explicit coupling scheme for poromechanics
Needed for benchmarking purposes, some draft is here https://github.com/GEOS-DEV/GEOS/pull/2646

### PR DESCRIPTION
**Pull request by paveltomin**
Has to go after https://github.com/GEOS-DEV/GEOS/pull/2906 and https://github.com/GEOS-DEV/GEOS/pull/2897

With changes in this PR explicit coupling seems to work fine and I am able to match FIM reference for SEAM case:
image
image
while simulation times:
```
SEQ           total time            05h44m08s (20648.672443899 s)
FIM           total time            1d, 07h38m18s (113898.859705466 s)
```

Also: force outer loop in poromechanics initialization

New baselines: https://github.com/GEOS-DEV/integratedTests/pull/111

### CODE REVIEW

File: src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp [hunk 1]
```diff
@@ -554,8 +555,9 @@ void TwoPointFluxApproximation::initNewFractureFieldsDFM( MeshLevel & mesh,
     }
     fluidPressure_n[newElemIndex] = fluidPressure[newElemIndex];
 #if ALLOW_CREATION_MASS==0
-    // set the initial density of the face element to 0 to enforce mass conservation ( i.e. no creation of mass)
-    dens[newElemIndex] = 0.0;
+    // set the initial mass/density of the face element to 0 to enforce mass conservation ( i.e. no creation of mass)
```
Comment (paveltomin) [inline src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp]: @CusiniM not sure about this piece, it does not seem even compile when ALLOW_CREATION_MASS is 0

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp [hunk 2]
```diff
@@ -244,6 +247,34 @@ void SinglePhaseBase::updateFluidModel( ObjectManagerBase & dataGroup ) const
...
+
+  SingleFluidBase & fluid =
+    getConstitutiveModel< SingleFluidBase >( subRegion, subRegion.getReference< string >( viewKeyStruct::fluidNamesString() ) );
+  arrayView2d< real64 const > const density = fluid.density();
+  arrayView2d< real64 const > const density_n = fluid.density_n();
+
+  forAll< parallelDevicePolicy<> >( subRegion.size(), [=] GEOS_HOST_DEVICE ( localIndex const ei )
+  {
+    mass[ei] = porosity[ei][0] * ( volume[ei] + deltaVolume[ei] ) * density[ei][0];
+    if( isZero( mass_n[ei] ) ) // this is a hack for hydrofrac cases
```
Comment (paveltomin) [inline src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:273]: @CusiniM this is also tricky, not sure how to properly do that

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.hpp
Comment (Guotong-Ren) [inline src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.hpp:1]: what is the change for here? is the modification to address the bulk modulus property in the first iteration of the iterative coupling?

File: src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp (hunk 1 above)
Comment (Guotong-Ren) [reply src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp]: i think in the fracture propagation, it provides different options to initialize the new element. At least in FP, new cells are always filled with fluid once generated.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.hpp
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.hpp:1]: Sorry, I don't see where are you pointing to?
The bulk modulus issue was addressed in https://github.com/GEOS-DEV/GEOS/pull/2766 and technically got merged into develop branch with https://github.com/GEOS-DEV/GEOS/pull/2568

**Review by jhuang2601 — APPROVED** [review]
It works well on SEAM CO2 case as expected!

File: src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp (hunk 1 above)
Comment (CusiniM) [reply src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp]: yes, it does offer different options but it was done through macros. This option was basically never used. This should be fixed in https://github.com/GEOS-DEV/GEOS/pull/2960 which we should review and merge.

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp (hunk 2 above)
Comment (CusiniM) [reply src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:273]: I think this should also be taken care in the initialization of new fractures. Why is the if statement necessary? isn't `mass_n` always equal to` porosity_n * volume_n * density_n` ?

File: src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp (hunk 1 above)
Comment (paveltomin) [reply src/coreComponents/finiteVolume/TwoPointFluxApproximation.cpp]: Is it ok if merge it like that for now and refactor after https://github.com/GEOS-DEV/GEOS/pull/2960 is merged?

File: src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp (hunk 2 above)
Comment (paveltomin) [reply src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp:273]: There is kind of cyclic dependency here https://github.com/GEOS-DEV/GEOS/blob/cb20b88c5e2d9afcf0b727bd8abb70f2c5366772/src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp#L618

- `mass_n` is saved from `mass` in `saveConvergedState`
- if I understand correctly `updatePorosityAndPermeability` needs `saveConvergedState` to be called before (see comment here https://github.com/GEOS-DEV/GEOS/blob/cb20b88c5e2d9afcf0b727bd8abb70f2c5366772/src/coreComponents/physicsSolvers/fluidFlow/SinglePhaseBase.cpp#L646)
- `updateMass` needs to be called the last to update `mass`

I am still not exactly sure how to untangle that...

File: src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp [hunk 3]
```diff
@@ -626,17 +631,17 @@ real64 CompositionalMultiphaseBase::updatePhaseVolumeFraction( ObjectManagerBase
   return maxDeltaPhaseVolFrac;
 }
 
-void CompositionalMultiphaseBase::updateFluidModel( ObjectManagerBase & dataGroup ) const
+void CompositionalMultiphaseBase::updateFluidModel( ElementSubRegionBase & subRegion ) const
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/fluidFlow/CompositionalMultiphaseBase.cpp]: I would have kept this an `ObjectManagerBase` though coz the method could work to update these fields on a `FaceManger` for example.

**Pull request merged**